[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127006
* #127960
* #127668
* __->__ #125382

Most commonly CPU scalars used for philox random seed. Right now, any cpu input will skip cudagraphing the entire graph. We need both the traced graph and the runtime inputs to be cudaified.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

xmfan committed:
[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -203,6 +207,38 @@ def post_acc_grad_hook(self, input, hook_id):
+            is_scalar = len(node.meta["val"].size()) == 0
+            if is_cpu and is_scalar:
+                # move cpu scalars to cuda in the graph
+                node.meta["val"] = node.meta["val"].cuda()
```
I think we need some additional checks here.  This code could run on a CPU-only graph and start moving scalars to cuda.  The machine might not even have cuda.

Perhaps check that at least one other input to this op is on cuda already.

eellison replied:
There is an existing pass to move constructors to cuda: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/post_grad.py#L916

We could add an option so that it moves input, cpu_scalars as well. Then we after applying, we could remove the conversions and take those as indices to convert. 

What point are we at the graph here ? I assume it's functionalized because backward. but is decomposed aten IR yet ?

xmfan replied:
> What point are we at the graph here ?

This is equivalent to user code. we call torch.compile on this GraphModule

> We could add an option so that it moves input, cpu_scalars as well

I'm not convinced we want to move graph inputs to cuda automatically. User may have code after calling the compiled function that still uses the inputs. In this case, the compiled autograd graph is the user code, and we have no need for these inputs after, so it is safe to move them.

eellison replied:
Do we know that all of the uses of the cpu_scalar dims are safe to use cuda tensors for? for instance if it is used in a custom op

eellison replied:
Hmm, maybe i'm not following but do we know that all of the uses of the cpu_scalar dims are safe to use cuda tensors for? for instance if it is used in a custom op. the other pass checks to see if all uses of a cpu constructor are aten/prims, and that all other inputs to a use can be moved as well

xmfan replied:
I added the aten/prims check on users of the cpu scalar we want to move. The cases are:

- python autograd.Function installed by AOTAutograd: CA traces into these, so all scalar users are visible, and scalar is moved iff ATen/prim.
- user-defined python autograd.Function: CA lifts these, so users are hidden. This will show up as a non-ATen/prim user to the scalar and the scalar will not be moved to CUDA.
- C++ autograd.Function: CA traces, scalar is moved iff ATen/prim.
- Backward hooks: Only gradients are inputs to backward hooks, so this case is affected by cpu scalar tensors that require grad. Since we have a check for CUDA inputs in the graph, we would also need a mixed device graph to trigger this case. And we don't seem to support this in fake tensor propagation, I added `test_scalar_grad_mixed_device` and xfail to track it.

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"


Most commonly CPU scalars used for philox random seed. Right now, any cpu input will skip cudagraphing the entire graph. We need both the traced graph and the runtime inputs to be cudaified.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"


Most commonly CPU scalars used for philox random seed. Right now, any cpu input will skip cudagraphing the entire graph. We need both the traced graph and the runtime inputs to be cudaified.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"


Most commonly CPU scalars used for philox random seed. Right now, any cpu input will skip cudagraphing the entire graph. We need both the traced graph and the runtime inputs to be cudaified.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd][cudagraphs] Inputs runtime wrapper to move cpu scalars to cuda"


Most commonly CPU scalars used for philox random seed. Right now, any cpu input will skip cudagraphing the entire graph. We need both the traced graph and the runtime inputs to be cudaified.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
